src/main/java/org/elasticsearch/cluster/action/index/NodeIndexDeletedAction.java

```diff
@@ -64,32 +75,56 @@ public void remove(Listener listener) {
     }
 
     public void nodeIndexDeleted(final ClusterState clusterState, final String index, final String nodeId) throws ElasticsearchException {
-        DiscoveryNodes nodes = clusterState.nodes();
+        final DiscoveryNodes nodes = clusterState.nodes();
         if (nodes.localNodeMaster()) {
-            threadPool.generic().execute(new Runnable() {
+            threadPool.generic().execute(new AbstractRunnable() {
+
+                @Override
+                public void onFailure(Throwable t) {
+                    logger.warn("failed to ack index store deleted for  index {}", t, index);
+                }
+
                 @Override
-                public void run() {
+                protected void doRun() throws Exception {
                     innerNodeIndexDeleted(index, nodeId);
+                    lockIndexAndAck(index, nodes, nodeId, clusterState);
+
                 }
             });
         } else {
             transportService.sendRequest(clusterState.nodes().masterNode(),
                     INDEX_DELETED_ACTION_NAME, new NodeIndexDeletedMessage(index, nodeId), EmptyTransportResponseHandler.INSTANCE_SAME);
+            threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
I agree with both of you. I'd like to keep things as they are semantically in this PR and I am happy to fix stuff in a different PR. This discussion might not be an easy one.